datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1356,16 +1392,82 @@ impl SMJStream {
                         pre_mask.clone()
                     };
 
+                    // Try to calculate if the buffered batch we scan is the last one for specific stream row and join key
+                    // for Batchsize == 1 self.buffered_data.scanning_finished() works well
+                    // For other scenarios its an attempt to figure out there is no more rows matching the same join key
+                    let last_batch = if self.batch_size == 1 {
```

Comment:
@korowa @viirya @alamb 
Hi guys appreciate if you have any other ideas how to calculate the last batch.
AntiJoin relies exactly on having the last batch, to calculate the predicate for join key correctly. 
 I'm trying to figure out there is no more buffered rows incoming for the given streaming join key. The approach is still not perfect as it still allows the tests to fail time to time although it becomes more stable.

Reply:
I will try and give it a look tomorrow

Reply:
I would expect the right check, based on function names, to be 
```rust
let last_batch = self.buffered_data.scanning_finished()
```

However, I tried that and the test still fails. 

@richox  I wonder if you have any ideas (as it appears you are the original author of SortMergeJoin in https://github.com/apache/datafusion/pull/2242)

I am having a hard time following the logic in such a large function (looks like `freeze_streamed` is something like `300` lines long).

If I were debugging this issue more, what I would probably do is
1.  to break the logic down into a few more named functions so the logic boundaries were clearer and the intended action is clearer. 
2. try and document, in comments, what the intended invariants of BufferedBatch / ScanningBatch are. My hope would be that in the process of writing that documentation I would learn the code more so I could have a better idea of what invariant isn't being upheld in this ufunction

Reply:
Documentation would def help. Btw here a ticket for SMJ documentation 
https://github.com/apache/datafusion/issues/10357

Reply:
I'm starting to look at this PR (will take some time though)

Reply:
> I'm starting to look at this PR (will take some time though)

Thanks, its not an actual PR, it is more attempts/directions to find a solution and discuss. Im experimenting more in parallel and would love to hear your ideas as well

Reply:
Thanks @korowa for the directions. This week I will try to find if such approach works for us, and alternatively I'm planning to play with a pair `scanning_batch().range.start` and `self.scanning_offset` perhaps it can give a hint how to identify last joined buffered side batch for the for the streaming row.

Reply:
Another option just came to my mind. We know in advance number of matched rows. We can calc it as 
```
let matched_indices = &self.buffered_data.batches.iter().map(|b| b.range.end - b.range.start).count();
```

No matter how SMJ distributes data, we can just take number of `buffered_indices.len()` per each iteration and substract it from `matched_indices`. Once we have hit 0 means no more matched rows expected